message.go

```diff
@@ -4,54 +4,246 @@ import (
 		}
+		r.err = errors.New(*errAsString)
+	}
+
+	// Decode the result
+	if r.c.res == nil {
+		r.c.res = new(interface{})
 	}
-	return nil
+	return d.Decode(r.c.res)
+}
+
+func (r RPCResponseData) SeqNo() seqNumber {
+	return r.c.seqid
 }
 
-func decodeMessage(dec decoder, m *message, i interface{}) error {
-	err := dec.Decode(i)
-	// TODO need to verify whether codec.Decode pulls from the reader if
-	// the decode fails, or whether the reader stays in the same state
-	// as before the Decode
-	m.remainingFields--
-	return err
+func (r RPCResponseData) Name() string {
+	return r.c.method
 }
 
-func decodeToNull(dec decoder, m *message) error {
-	var err error
-	for err == nil && m.remainingFields > 0 {
-		i := new(interface{})
-		err = decodeMessage(dec, m, i)
+func (r RPCResponseData) Err() error {
+	return r.err
+}
+
+func (r RPCResponseData) Res() interface{} {
+	return r.c.res
+}
+
+func (r RPCResponseData) ResponseCh() chan RPCMessage {
+	return r.c.resultCh
+}
+
+type RPCNotifyData struct {
+	BasicRPCData
+	name string
+	arg  interface{}
+}
+
+func (r *RPCNotifyData) DecodeData(l int, d decoder, p *protocolHandler, _ *callContainer) (err error) {
+	if err = d.Decode(&r.name); err != nil {
+		return err
 	}
-	return err
+	if r.arg, err = p.getArg(r.name); err != nil {
+		return err
+	}
+	return d.Decode(r.arg)
 }
 
-func decodeError(dec decoder, m *message, f ErrorUnwrapper) (appErr error, dispatchErr error) {
-	if f != nil {
-		arg := f.MakeArg()
-		err := decodeMessage(dec, m, arg)
-		if err != nil {
-			return nil, err
-		}
-		return f.UnwrapError(arg)
+func (RPCNotifyData) MinLength() int {
+	return 2
+}
+
+func (r RPCNotifyData) SeqNo() seqNumber {
+	return -1
+}
+
+func (r RPCNotifyData) Name() string {
+	return r.name
+}
+
+func (r RPCNotifyData) Arg() interface{} {
+	return r.arg
+}
+
+type RPCCancelData struct {
+	BasicRPCData
+	seqno seqNumber
+	name  string
+}
+
+func (r *RPCCancelData) DecodeData(l int, d decoder, p *protocolHandler, _ *callContainer) (err error) {
+	if err = d.Decode(&r.seqno); err != nil {
+		return err
+	}
+	return d.Decode(&r.name)
+}
+
+func (RPCCancelData) MinLength() int {
+	return 2
+}
+
+func (r RPCCancelData) SeqNo() seqNumber {
+	return r.seqno
+}
+
+func (r RPCCancelData) Name() string {
+	return r.name
+}
+
+type RPCCall struct {
+	typ MethodType
+
+	RPCData
+}
+
+func (c RPCCall) Type() MethodType {
+	return c.typ
+}
+
+func (c *RPCCall) Decode(l int, d decoder, p *protocolHandler, cc *callContainer) error {
+	if err := d.Decode(&c.typ); err != nil {
+		return err
+	}
+
+	switch c.typ {
+	case MethodCall:
+		c.RPCData = &RPCCallData{}
+	case MethodResponse:
+		c.RPCData = &RPCResponseData{}
+	case MethodNotify:
+		c.RPCData = &RPCNotifyData{}
+	case MethodCancel:
+		c.RPCData = &RPCCancelData{}
+	default:
+		c.RPCData = nil
+		return errors.New("invalid RPC type")
 	}
-	var s string
-	if dispatchErr = decodeMessage(dec, m, &s); dispatchErr == nil && len(s) > 0 {
-		appErr = errors.New(s)
+	l--
```

Comment:
I feel like this check (modified appropriately) should go first, so we don't even attempt to decode if the length is bad